arrow-fx/src/main/kotlin/arrow/fx/internal/IOBracket.kt

```diff
@@ -87,86 +95,62 @@ internal object IOBracket {
       }
     }
 
-  private class BracketReleaseFrame<E, A, B>(val a: A, val releaseFn: (A, ExitCase2<E>) -> IOOf<E, Unit>) :
-    BaseReleaseFrame<E, A, B>() {
+  private class BracketReleaseFrame<A, B>(val a: A, val releaseFn: (A, ExitCase<Throwable>) -> IOOf<Unit>) :
+    BaseReleaseFrame<A, B>() {
 
-    override fun release(c: ExitCase2<E>): IOOf<E, Unit> =
+    override fun release(c: ExitCase<Throwable>): CancelToken<ForIO> =
       releaseFn(a, c)
   }
 
-  private class EnsureReleaseFrame<E, A>(val releaseFn: (ExitCase2<E>) -> IOOf<E, Unit>) : BaseReleaseFrame<E, Unit, A>() {
+  private class EnsureReleaseFrame<A>(val releaseFn: (ExitCase<Throwable>) -> IOOf<Unit>) : BaseReleaseFrame<Unit, A>() {
 
-    override fun release(c: ExitCase2<E>): IOOf<E, Unit> = releaseFn(c)
+    override fun release(c: ExitCase<Throwable>): CancelToken<ForIO> = releaseFn(c)
   }
 
-  private abstract class BaseReleaseFrame<E, A, B> : IOFrame<E, B, IO<E, B>> {
+  private abstract class BaseReleaseFrame<A, B> : IOFrame<B, IO<B>> {
 
     // Guard used for thread-safety, to ensure the idempotency
     // of the release; otherwise `release` can be called twice
     private val waitsForResult = atomic(true)
 
-    abstract fun release(c: ExitCase2<E>): IOOf<E, Unit>
+    abstract fun release(c: ExitCase<Throwable>): CancelToken<ForIO>
 
-    private fun applyRelease(e: ExitCase2<E>): IO<E, Unit> =
+    private fun applyRelease(e: ExitCase<Throwable>): IO<Unit> =
       IO.defer {
         if (waitsForResult.compareAndSet(true, false)) release(e)
         else IO.unit
       }
 
-    val cancel: IOOf<E, Unit> = applyRelease(ExitCase2.Cancelled).fix().uncancellable()
+    val cancel: CancelToken<ForIO> = applyRelease(ExitCase.Cancelled).fix().uncancellable()
 
     // Unregistering cancel token, otherwise we can have a memory leak;
     // N.B. conn.pop() happens after the evaluation of `release`, because
     // otherwise we might have a conflict with the auto-cancellation logic
-    override fun recover(e: Throwable): IO<E, B> =
-      IO.Bind(IO.ContextSwitch(applyRelease(ExitCase2.Exception(e)), IO.ContextSwitch.makeUncancellable, disableUncancellableAndPop),
-        ReleaseRecoverException(e))
+    override fun recover(e: Throwable): IO<B> =
+      IO.Bind(
```

Comment:
@nomisRev I brought back this as well FYI

Reply:
👌